File: CS/Reporting-How-To-Use-Object-Data-Source/WebForms/VB/Reporting_ObjectDS_WebForms/Services/CustomReportStorageWebExtension.vb
Imports System
Imports System.Collections.Concurrent
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Reflection
Imports System.ServiceModel
Imports DevExpress.XtraReports.Web.Extensions
Imports DevExpress.XtraReports.UI
Imports Reporting_ObjectDS_WebForms.PredefinedReports

Namespace Reporting_ObjectDS_WebForms.Services
	Public Class CustomReportStorageWebExtension
		Inherits DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension

		Private ReadOnly reportDirectory As String
		Private Const FileExtension As String = ".repx"
		Public Sub New(ByVal reportDirectory As String)
			If Not Directory.Exists(reportDirectory) Then
				Directory.CreateDirectory(reportDirectory)
			End If
			Me.reportDirectory = reportDirectory
		End Sub

		Private Function IsWithinReportsFolder(ByVal url As String, ByVal folder As String) As Boolean
			Dim rootDirectory = New DirectoryInfo(folder)
			Dim fileInfo = New FileInfo(Path.Combine(folder, url))
			Return fileInfo.Directory.FullName.ToLower().StartsWith(rootDirectory.FullName.ToLower())
		End Function

		Public Overrides Function CanSetData(ByVal url As String) As Boolean
			' Determines whether or not it is possible to store a report by a given URL. 
			' For instance, make the CanSetData method return false for reports that should be read-only in your storage. 
			' This method is called only for valid URLs (i.e., if the IsValidUrl method returned true) before the SetData method is called.

			Return True
		End Function

		Public Overrides Function IsValidUrl(ByVal url As String) As Boolean
			' Determines whether or not the URL passed to the current Report Storage is valid. 
			' For instance, implement your own logic to prohibit URLs that contain white spaces or some other special characters. 
			' This method is called before the CanSetData and GetData methods.

			Return Path.GetFileName(url) = url
		End Function

		Public Overrides Function GetData(ByVal url As String) As Byte()
			' Returns report layout data stored in a Report Storage using the specified URL. 
			' This method is called only for valid URLs after the IsValidUrl method is called.
			Try
				If Directory.EnumerateFiles(reportDirectory).Select(AddressOf Path.GetFileNameWithoutExtension).Contains(url) Then
					Return File.ReadAllBytes(Path.Combine(reportDirectory, url & FileExtension))
				End If
				If ReportsFactory.Reports.ContainsKey(url) Then
					Using ms As New MemoryStream()
						ReportsFactory.Reports(url)().SaveLayoutToXml(ms)
						Return ms.ToArray()
					End Using
				End If
			Catch e1 As Exception
				Throw New FaultException(New FaultReason("Could not get report data."), New FaultCode("Server"), "GetData")
			End Try
			Throw New FaultException(New FaultReason(String.Format("Could not find report '{0}'.", url)), New FaultCode("Server"), "GetData")
		End Function

		Public Overrides Function GetUrls() As Dictionary(Of String, String)
			' Returns a dictionary of the existing report URLs and display names. 
			' This method is called when running the Report Designer, 
			' before the Open Report and Save Report dialogs are shown and after a new report is saved to a storage.

			Return Directory.GetFiles(reportDirectory, "*" & FileExtension).Select(AddressOf Path.GetFileNameWithoutExtension).Union(ReportsFactory.Reports.Select(Function(x) x.Key)).ToDictionary(Function(x) x)
		End Function

		Public Overrides Sub SetData(ByVal report As XtraReport, ByVal url As String)
			' Stores the specified report to a Report Storage using the specified URL. 
			' This method is called only after the IsValidUrl and CanSetData methods are called.
			If Not IsWithinReportsFolder(url, reportDirectory) Then
				Throw New FaultException(New FaultReason("Invalid report name."), New FaultCode("Server"), "GetData")
			End If
			report.SaveLayoutToXml(Path.Combine(reportDirectory, url & FileExtension))
		End Sub

		Public Overrides Function SetNewData(ByVal report As XtraReport, ByVal defaultUrl As String) As String
			' Stores the specified report using a new URL. 
			' The IsValidUrl and CanSetData methods are never called before this method. 
			' You can validate and correct the specified URL directly in the SetNewData method implementation 
			' and return the resulting URL used to save a report in your storage.
			SetData(report, defaultUrl)
			Return defaultUrl
		End Function
	End Class
End Namespace

File: MVC/CS/Reporting_ObjectDS_Mvc/Services/CustomReportStorageWebExtension.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.ServiceModel;
using DevExpress.XtraReports.Web.Extensions;
using DevExpress.XtraReports.UI;
using Reporting_ObjectDS_Mvc.PredefinedReports;

namespace Reporting_ObjectDS_Mvc.Services
{
    public class CustomReportStorageWebExtension : DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension
    {
        readonly string reportDirectory;
        const string FileExtension = ".repx";
        public CustomReportStorageWebExtension(string reportDirectory) {
            if (!Directory.Exists(reportDirectory)) {
                Directory.CreateDirectory(reportDirectory);
            }
            this.reportDirectory = reportDirectory;
        }

        private bool IsWithinReportsFolder(string url, string folder) {
            var rootDirectory = new DirectoryInfo(folder);
            var fileInfo = new FileInfo(Path.Combine(folder, url));
            return fileInfo.Directory.FullName.ToLower().StartsWith(rootDirectory.FullName.ToLower());
        }

        public override bool CanSetData(string url) {
            // Determines whether or not it is possible to store a report by a given URL. 
            // For instance, make the CanSetData method return false for reports that should be read-only in your storage. 
            // This method is called only for valid URLs (i.e., if the IsValidUrl method returned true) before the SetData method is called.

            return true;
        }

        public override bool IsValidUrl(string url) {
            // Determines whether or not the URL passed to the current Report Storage is valid. 
            // For instance, implement your own logic to prohibit URLs that contain white spaces or some other special characters. 
            // This method is called before the CanSetData and GetData methods.

            return Path.GetFileName(url) == url;
        }

        public override byte[] GetData(string url) {
            // Returns report layout data stored in a Report Storage using the specified URL. 
            // This method is called only for valid URLs after the IsValidUrl method is called.
            try {
                if (Directory.EnumerateFiles(reportDirectory).Select(Path.GetFileNameWithoutExtension).Contains(url))
                {
                    return File.ReadAllBytes(Path.Combine(reportDirectory, url + FileExtension));
                }
                if (ReportsFactory.Reports.ContainsKey(url))
                {
                    using (MemoryStream ms = new MemoryStream()) {
                        ReportsFactory.Reports[url]().SaveLayoutToXml(ms);
                        return ms.ToArray();
                    }
                }
            } catch (Exception) {
                throw new FaultException(new FaultReason("Could not get report data."), new FaultCode("Server"), "GetData");
            }
            throw new FaultException(new FaultReason(string.Format("Could not find report '{0}'.", url)), new FaultCode("Server"), "GetData");
        }

        public override Dictionary<string, string> GetUrls() {
            // Returns a dictionary of the existing report URLs and display names. 
            // This method is called when running the Report Designer, 
            // before the Open Report and Save Report dialogs are shown and after a new report is saved to a storage.
            
            return Directory.GetFiles(reportDirectory, "*" + FileExtension)
                                     .Select(Path.GetFileNameWithoutExtension)
                                     .Union(ReportsFactory.Reports.Select(x => x.Key))
                                     .ToDictionary<string, string>(x => x);
        }

        public override void SetData(XtraReport report, string url) {
            // Stores the specified report to a Report Storage using the specified URL. 
            // This method is called only after the IsValidUrl and CanSetData methods are called.
            if(!IsWithinReportsFolder(url, reportDirectory))
                throw new FaultException(new FaultReason("Invalid report name."), new FaultCode("Server"), "GetData");
            report.SaveLayoutToXml(Path.Combine(reportDirectory, url + FileExtension));
        }

        public override string SetNewData(XtraReport report, string defaultUrl) {
            // Stores the specified report using a new URL. 
            // The IsValidUrl and CanSetData methods are never called before this method. 
            // You can validate and correct the specified URL directly in the SetNewData method implementation 
            // and return the resulting URL used to save a report in your storage.
            SetData(report, defaultUrl);
            return defaultUrl;
        }
    }
}
